cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/GoalViolationDetector.java

```diff
@@ -87,68 +86,49 @@ public GoalViolationDetector(KafkaCruiseControlConfig config,
-      LoadMonitorTaskRunner.LoadMonitorTaskRunnerState loadMonitorTaskRunnerState = _loadMonitor.taskRunnerState();
-      if (!ViolationUtils.isLoadMonitorReady(loadMonitorTaskRunnerState)) {
-        LOG.info("Skipping goal violation detection because load monitor is in {} state.", loadMonitorTaskRunnerState);
-        return;
-      }
+    LoadMonitorTaskRunner.LoadMonitorTaskRunnerState loadMonitorTaskRunnerState = _loadMonitor.taskRunnerState();
+    if (!ViolationUtils.isLoadMonitorReady(loadMonitorTaskRunnerState)) {
+      LOG.info("Skipping goal violation detection because load monitor is in {} state.", loadMonitorTaskRunnerState);
+      return;
+    }
 
+    List<Goal> readyGoals =  _goals.values().stream()
+                                   .filter(goal -> _loadMonitor.meetCompletenessRequirements(goal.clusterModelCompletenessRequirements()))
+                                   .collect(Collectors.toList());
+    if (!readyGoals.isEmpty()) {
+      LOG.debug("Detecting if ready goals {} out of {} are violated.", readyGoals, _goals.values());
       GoalViolations goalViolations = new GoalViolations(_kafkaCruiseControl, _allowCapacityEstimation);
-      boolean newModelNeeded = true;
-      ClusterModel clusterModel = null;
-      for (Map.Entry<Integer, Goal> entry : _goals.entrySet()) {
-        Goal goal = entry.getValue();
-        if (_loadMonitor.meetCompletenessRequirements(goal.clusterModelCompletenessRequirements())) {
-          LOG.debug("Detecting if {} is violated.", entry.getValue().name());
-          // Because the model generation could be slow, We only get new cluster model if needed.
-          if (newModelNeeded) {
-            if (clusterModelSemaphore != null) {
-              clusterModelSemaphore.close();
-            }
-            clusterModelSemaphore = _loadMonitor.acquireForModelGeneration(new OperationProgress());
-            // Make cluster model null before generating a new cluster model so the current one can be GCed.
-            clusterModel = null;
-            clusterModel = _loadMonitor.clusterModel(now,
-                                                     goal.clusterModelCompletenessRequirements(),
-                                                     new OperationProgress());
-            KafkaCruiseControl.sanityCheckCapacityEstimation(_allowCapacityEstimation,
-                                                             clusterModel.capacityEstimationInfoByBrokerId());
-          }
-          int priority = entry.getKey();
-          newModelNeeded = optimizeForGoal(clusterModel, priority, goal, goalViolations);
-        } else {
-          LOG.debug("Skipping goal violation detection for {} because load completeness requirement is not met.", goal);
-        }
+      ModelCompletenessRequirements requirements = new ModelCompletenessRequirements(0, 0, false);
+      for (Goal goal: readyGoals) {
+        requirements = requirements.stronger(goal.clusterModelCompletenessRequirements());
```

Comment:
What is the reasoning for using combined requirements?

IIUC, contrary to before, due to this stronger requirements, the goal violation detector would fail to detect violations of goals with weak requirements (e.g. `rack awareness`) due to the stronger requirements of some other goals -- such as `CpuUsageDistributionGoal` because the monitored cluster model does not have the required number of valid windows, which has nothing to do with detecting `rack awareness`.

I think we should ensure that we do not let the ability to fix goal violations impact the ability to detect goal violations.

Reply:
My original purpose is to reduce the number of times calculating `ClusterModel`, which is a source-heavy thing. My initial idea is to get a proper requirement and just generate cluster model once, but you are right, even `LoadMonitor` can meet each ready goal, it may not meet the `stronger()` of all goals.

Reply:
And also I do not understand the origin code's logic around `NewModelNeeded`.

Reply:
on a second observation. for all the goals we currently support. 
RackAwareGoal, (requiredNumWindows=1, minMonitoredPartitionPercentage=0.000, includedAllTopics=true)
ReplicaCapacityGoal, (requiredNumWindows=1, minMonitoredPartitionPercentage=0.000, includedAllTopics=true)
DiskCapacityGoal, (requiredNumWindows=1, minMonitoredPartitionPercentage=0.980, includedAllTopics=true)
NetworkInboundCapacityGoal, (requiredNumWindows=1, minMonitoredPartitionPercentage=0.980, includedAllTopics=true)
NetworkOutboundCapacityGoal, (requiredNumWindows=1, minMonitoredPartitionPercentage=0.980, includedAllTopics=true)
CpuCapacityGoal, (requiredNumWindows=1, minMonitoredPartitionPercentage=0.980, includedAllTopics=true)
ReplicaDistributionGoal, (requiredNumWindows=1, minMonitoredPartitionPercentage=0.000, includedAllTopics=true)
PotentialNwOutGoal, (requiredNumWindows=168, minMonitoredPartitionPercentage=0.980, includedAllTopics=false)
DiskUsageDistributionGoal, (requiredNumWindows=1, minMonitoredPartitionPercentage=0.980, includedAllTopics=true)
NetworkInboundUsageDistributionGoal, (requiredNumWindows=168, minMonitoredPartitionPercentage=0.980, includedAllTopics=false)
NetworkOutboundUsageDistributionGoal, (requiredNumWindows=168, minMonitoredPartitionPercentage=0.980, includedAllTopics=false)
CpuUsageDistributionGoal, (requiredNumWindows=168, minMonitoredPartitionPercentage=0.980, includedAllTopics=false)
TopicReplicaDistributionGoal, (requiredNumWindows=1, minMonitoredPartitionPercentage=0.000, includedAllTopics=true)
LeaderBytesInDistributionGoal, (requiredNumWindows=168, minMonitoredPartitionPercentage=0.980, includedAllTopics=false)

there are actually 3 types of complete requirement 
1. window=1, partition_percentage=0, all_topic=true
2.window=1, partition_percentage=0.98, all_topic=true
3.window=1, partition_percentage=0.98, all_topic=false

looks like `stronger()` of all ready goal is guaranteed to meet. we do not have the issue of `impact the ability to detect goal violations`.

Reply:
Thanks for the analysis! -- here are some thoughts:

1. While I agree that redundant calculations of `clusterModel` should be avoided, I am not entirely sure if the current implementation of this part had redundant calculations of cluster model. Note that the current implementation calculates the cluster model only if the cluster had changes (e.g. `if (newModelNeeded) == true`). If we skip this step and use an old cluster model, then we potentially would have `false positives` or `false negatives`. Please note that `GoalViolationDetector#optimizeForGoal` modifies the passed `clusterModel` if there has been any violations to the goal. Using this modified cluster model does not provide an accurate picture of if a goal has been violated or not.

2. (potential future improvement) In the current implementation, we keep a single `clusterModel`, and update the model when/if necessary. But there are some goals which only deal with the topology of the cluster model. For such goals, we can work with a "skinnier" version of the cluster model, which does not need to perform an load aggregation. This would, I presume, provide improvements in the time it takes to generate a cluster model.

Let me know if I can provide further clarification!

Reply:
I missed the point that `GoalViolationDetector#optimizeForGoal` will modify the `ClusterModel`(no wonder the original code logic looks confusing to me...). then in this case what do you feel about adding a `clone()` method to `ClusterModel`(for the sake of reducing the times of  load aggregation)?